core/consensus/common/src/import_queue.rs

```diff
@@ -70,20 +72,17 @@ pub trait Verifier<B: BlockT>: Send + Sync + Sized {
 		origin: BlockOrigin,
 		header: B::Header,
 		justification: Option<Justification>,
-		body: Option<Vec<B::Extrinsic>>
+		body: Option<Vec<B::Extrinsic>>,
 	) -> Result<(ImportBlock<B>, Option<Vec<AuthorityIdFor<B>>>), String>;
 }
 
 /// Blocks import queue API.
-pub trait ImportQueue<B: BlockT>: Send + Sync {
+pub trait ImportQueue<B: BlockT>: Send + Sync + ImportQueueClone<B> {
 	/// Start background work for the queue as necessary.
 	///
 	/// This is called automatically by the network service when synchronization
 	/// begins.
-	fn start<L>(&self, _link: L) -> Result<(), std::io::Error> where
-		Self: Sized,
-		L: 'static + Link<B>,
-	{
+	fn start(&self, _link: Box<Link<B>>) -> Result<(), std::io::Error> {
```

Comment:
Ok so I've tried creating the workers lazily when starting, the problem with that is that you end-up trading a single box for 2-3 additional stuff stored on the `BasicQueue`, since those things are passed-in when `new` is called, which is in a different place from where `start` is called. I could do something like storing a bunch of `Option` and then `take` the stuff out of them, I guess, and that feels quite complicated. 

After that, another problem I ran in was that what was previously done with passing `Link` around as a generic param in the form of `start<L: 'static + Link<B>>`, actually requires the `start` method on the `ImportQueue` trait to be generic, and we can't `Box` a trait with generic methods...

The reason why `ImportQueue` is passed around in `Box`, is because it's now a stateless wrapper around a sender, while previously it would be `Arc<ImportQueue>` since the struct that implemented `ImportQueue` also contained all the state related to the implementation( see https://github.com/paritytech/substrate/blob/c188a3dc1505167748121a8f6cb9e970ff47df91/core/consensus/common/src/import_queue.rs#L113). 

I'm not very familiar with the cost tradeoffs between `Arc` and `Box`, and my feeling is that the use of box makes sense, and we're gaining because we're not sharing state anymore. If that costs us an additional box for the `Link`, which previously was a generic param, it might be worth it. 

What do you think?

Reply:
I'm also not very familiar with those cost tradeoffs, but I agree with your points. The only thing I don't like about the worker is really the exceptional "not started" state, I think an option would be to make it kind of a "builder pattern", so that `start` would return a new type (which just wraps the sender). But I'm not sure that would work and I don't have a strong opinion on this, I'm fine with logging the exceptional case.